systemd: add openssl as explicit buildInput

Shawn8901 opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

fixes #369446

openssl is propagated by libfido2, which breaks systemd compliation in case fido support is disabled.

importd should enable openssl if gcrypt is disabled https://github.com/systemd/systemd/blob/89c4fe6c211d5e9cb380329f638609dec26bed0a/meson.build#L1637


I am still building, tho i already wanted to create the pr to gather feedback from the issue creator. Aside already that every feedback is welcome.




## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Shawn8901 committed:
systemd: add openssl as explicit buildInput

ambroisie commented:
The condition for `openssl` is quite complicated IMO, could it get factorized to a single place (either a local variable, or a new `with*` option, in the latter case with assertions for incompatible options)?

Shawn8901 commented:
> or a new `with*` option, in the latter case with assertions for incompatible options
So that we assert, that when withHomed is true, we need to have set withOpenSSL?

ambroisie commented:
If I understand correctly, something like:

```nix
assert withHomed -> withOpenssl;
assert withFido2 -> withOpenssl;
assert withSysupdate -> withOpenssl;
assert withImportd -> (wantGcrypt || withOpenssl);
assert withOpenssl -> !wantGcrypt;
```
Or something like that.

ambroisie commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -178,13 +181,17 @@ assert withImportd -> withCompression;
 
 let
   wantCurl = withRemote || withImportd;
-  wantGcrypt = withResolved || withImportd;
```
Should also add  a  `withResolved -> withGcrypt` assertion?

Shawn8901 replied:
see the comment below, sorry my internet was down, so the comment i wanted to add, did take some time :slightly_smiling_face:

Shawn8901 replied:
> Are withGcrypt = true and withOpenssl = true compatible?

explicit yes, as that was also already possible on the old setup (withFido did include openssl and resolved enabled gcrypt, which are both active on default)

ambroisie replied:
If it can build without it, I think it's fine to leave it out. Assertions are added to make it easier to debug build failures (rather than attempt to build and have the user dig through the logs to find out their options aren't compatible).

ambroisie reviewed (commented):
Are `withGcrypt = true` and `withOpenssl = true` compatible?

Shawn8901 commented:
Applied that refactor.

While checking the meson.build i did not find any evidence that resolved actually requires gcrypt ( https://github.com/systemd/systemd/blob/89c4fe6c211d5e9cb380329f638609dec26bed0a/meson.build#L3110 ), tho i have removed that.
Importd supports both openssl or gcrypt according the meson.build  ( https://github.com/systemd/systemd/blob/89c4fe6c211d5e9cb380329f638609dec26bed0a/meson.build#L1639 )

So i switched the gcrypt to a withGcrypt to comply with the other options.

Shawn8901 commented:
i has build my system with that and the over-written config from the related issue (minus coredump support, which i kept enabled). Seeing no issue with my build, so i am undrafting.

ElvishJerricco reviewed (changes requested):
I'm pretty sure `systemdMinimal` needs to set `withOpenssl = false;` and `withGcrypt = false;`, right?

Shawn8901 commented:
> I'm pretty sure `systemdMinimal` needs to set `withOpenssl = false;` and `withGcrypt = false;`, right?

seems absolutely reasonable. amended that into the commit.
